File: training-data/perl/rod_cutting.pl
use strict;
use warnings;

sub rod_cutting {
    my ($prices, $length) = @_;
    my @dp = (0) x ($length + 1);
    for my $len (1 .. $length) {
        my $best = -1;
        for my $cut (1 .. $len) {
            next unless defined $prices->[$cut - 1];
            my $value = $prices->[$cut - 1] + $dp[$len - $cut];
            $best = $value if $value > $best;
        }
        $dp[$len] = $best;
    }
    return $dp[$length];
}

my @prices = (1, 5, 8, 9, 10, 17, 17, 20);
print rod_cutting(\@prices, 8), "\n";

File: training-data/kotlin/RodCutting.kt
fun rodCutting(prices: IntArray, length: Int): Int {
    val dp = IntArray(length + 1)
    for (i in 1..length) {
        var best = -1
        for (cut in 1..i) {
            if (cut <= prices.size) {
                val value = prices[cut - 1] + dp[i - cut]
                if (value > best) best = value
            }
        }
        dp[i] = best
    }
    return dp[length]
}

fun main() {
    val prices = intArrayOf(1, 5, 8, 9, 10, 17, 17, 20)
    for (length in 1..prices.size) {
        println("length $length -> best value ${rodCutting(prices, length)}")
    }
}
